### PR TITLE
Go back to the previous ABBP algo and fix it

### PR DESCRIPTION
**Pull request by lorban**
Alternative to #11426

### CODE REVIEW

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java [hunk 1]
```diff
@@ -415,14 +330,13 @@ public long getHeapMemory()
...
-
-    private void subtractMemory(int amount, boolean direct)
-    {
-        AtomicLong memory = direct ? _directMemory : _heapMemory;
-        memory.addAndGet(-amount);
+        RetainedBucket[] buckets = direct ? _direct : _indirect;
+        long total = 0L;
+        for (RetainedBucket bucket : buckets)
+        {
+            total += bucket._pool.size() * (long)bucket._capacity;
```
Comment (gregw) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:337]: I'm not sure what this number is?    It is all idle and inuse buffers from any concurrent pool plus just the idle buffers from any queued pool.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java [hunk 2]
```diff
@@ -466,15 +377,84 @@ private void clearBuckets(RetainedBucket[] buckets)
...
-        Dumpable.dumpObjects(
-            out,
-            indent,
-            this,
-            DumpableCollection.fromArray("direct", _direct),
-            DumpableCollection.fromArray("indirect", _indirect));
+        long maxMemory = direct ? _maxDirectMemory : _maxHeapMemory;
+        if (maxMemory > 0)
+        {
+            long excess = getMemory(direct) - maxMemory;
```
Comment (gregw) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:385]: This is a full iteration over buckets with a full iteration over each concurrent pool entry for every pool miss.       This is not different to my PRs always checking memory in the retain method.

However, you do not check max memory for a normal release, and it certainly can be exceeded (especially with the strange impl of getMemory).

**Comment by lorban** [comment]
I'm not convinced this implementation is better than https://github.com/jetty/jetty.project/pull/11426 as when it reaches memory capacity, very many acquisitions won't be served and will trigger an eviction, making the performance plummet quite badly.

Here's what the microbenchmark has to say about this:

```
Benchmark                                                                     (factor)  (maxBucketSize)  (maxCapacity)  (maxMemory)  (minCapacity)  (type)   Mode  Cnt          Score          Error  Units
ArrayByteBufferPoolBenchmark.fastPathAcquireRelease                               4096               -1          65536            0              0    greg  thrpt   10  229142102.162 ± 35233219.736  ops/s
ArrayByteBufferPoolBenchmark.fastPathAcquireRelease                               4096               -1          65536            0              0    ludo  thrpt   10  381324499.460 ±  8712489.462  ops/s
ArrayByteBufferPoolBenchmark.fastPathAcquireRelease                               4096               -1          65536      1048576              0    greg  thrpt   10  215758553.527 ±  8279594.160  ops/s
ArrayByteBufferPoolBenchmark.fastPathAcquireRelease                               4096               -1          65536      1048576              0    ludo  thrpt   10  394587715.082 ±  4493314.070  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacity               4096               -1          65536            0              0    greg  thrpt   10   81966597.179 ±  2779241.727  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacity               4096               -1          65536            0              0    ludo  thrpt   10   78507283.615 ±  8224128.776  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacity               4096               -1          65536      1048576              0    greg  thrpt   10   27888036.667 ±  9597693.594  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacity               4096               -1          65536      1048576              0    ludo  thrpt   10     309518.284 ±    65010.749  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacityMigrating      4096               -1          65536            0              0    greg  thrpt   10   79412953.878 ±  4868374.327  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacityMigrating      4096               -1          65536            0              0    ludo  thrpt   10   84964065.875 ± 11437317.943  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacityMigrating      4096               -1          65536      1048576              0    greg  thrpt   10   25604116.168 ± 40112738.478  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacityMigrating      4096               -1          65536      1048576              0    ludo  thrpt   10   30624600.597 ± 47674548.825  ops/s
```

**Comment by lorban** [comment]
Here's what the microbenchmark has to say about this:

```
Benchmark                                                                     (factor)  (maxBucketSize)  (maxCapacity)  (maxMemory)  (minCapacity)  (type)   Mode  Cnt          Score          Error  Units
ArrayByteBufferPoolBenchmark.fastPathAcquireRelease                               4096               -1          65536            0              0    greg  thrpt   10  229142102.162 ± 35233219.736  ops/s
ArrayByteBufferPoolBenchmark.fastPathAcquireRelease                               4096               -1          65536            0              0    ludo  thrpt   10  381324499.460 ±  8712489.462  ops/s
ArrayByteBufferPoolBenchmark.fastPathAcquireRelease                               4096               -1          65536      1048576              0    greg  thrpt   10  215758553.527 ±  8279594.160  ops/s
ArrayByteBufferPoolBenchmark.fastPathAcquireRelease                               4096               -1          65536      1048576              0    ludo  thrpt   10  394587715.082 ±  4493314.070  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacity               4096               -1          65536            0              0    greg  thrpt   10   81966597.179 ±  2779241.727  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacity               4096               -1          65536            0              0    ludo  thrpt   10   78507283.615 ±  8224128.776  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacity               4096               -1          65536      1048576              0    greg  thrpt   10   27888036.667 ±  9597693.594  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacity               4096               -1          65536      1048576              0    ludo  thrpt   10     309518.284 ±    65010.749  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacityMigrating      4096               -1          65536            0              0    greg  thrpt   10   79412953.878 ±  4868374.327  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacityMigrating      4096               -1          65536            0              0    ludo  thrpt   10   84964065.875 ± 11437317.943  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacityMigrating      4096               -1          65536      1048576              0    greg  thrpt   10   25604116.168 ± 40112738.478  ops/s
ArrayByteBufferPoolBenchmark.inputFixedCapacityOutputRandomCapacityMigrating      4096               -1          65536      1048576              0    ludo  thrpt   10   30624600.597 ± 47674548.825  ops/s
```

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java (hunk 2 above)
Comment (lorban) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:385]: It's worse than your impl, and the memory accounting is quite strange indeed.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java (hunk 1 above)
Comment (lorban) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:337]: It's the known memory consumption, which doesn't mean much.

**Comment by lorban** [comment]
It was worth doing for comparison, but the alternative in https://github.com/jetty/jetty.project/pull/11426 is better.

**Pull request closed without merging**